core/unix/memquery_test.h

```diff
@@ -0,0 +1,311 @@
+                                 OUT app_pc *out_max_end /* relative pc */,
+                                 OUT char **out_soname, OUT os_module_data_t *out_data)
+{
+    if (!enable_memquery_unit_test) {
+        return module_walk_program_headers(
+            base, view_size, at_map, dyn_reloc, out_base /* relative pc */,
+            out_first_end /* relative pc */, out_max_end /* relative pc */, out_soname,
+            out_data);
+    }
+    if (cur_bounds.iters_count == 0) {
+        ASSERT(false && "module_is_header_test called without setting cur_bounds!");
+    }
+    if (out_first_end != NULL || out_soname != NULL || out_data != NULL ||
+        out_base == NULL || out_max_end == NULL) {
+        ASSERT(false &&
+               "out_data, out_first_end, and out_soname must be NULL, and out_base and "
+               "out_max_end must not be!");
+    }
+    for (int i = 0; i < cur_bounds.iters_count; i++) {
+        fake_memquery_result *cur = &cur_bounds.iters[i];
+        memquery_iter_t iter = cur->iter_result;
+        if (iter.vm_start <= base && base < iter.vm_end) {
+            if (cur->is_header) {
+                *out_base = cur->mod_base;
+                *out_max_end = cur->mod_end;
+                return true;
+            }
+        }
+    }
+    ASSERT(false && "UNKNOWN BASE PC");
+    return false;
+}
+
+/* record_memquery_library_bounds_by_iterator wraps
+ * memquery_library_bounds_by_iterator, printing out test case data for the
+ * memquery unit test. See memquery_test_data.h for a selection of tests.
+ */
+int
+record_memquery_library_bounds_by_iterator(const char *name, app_pc *start /*IN/OUT*/,
+                                           app_pc *end /*OUT*/,
+                                           char *fulldir /*OPTIONAL OUT*/,
+                                           size_t fulldir_size,
+                                           char *filename /*OPTIONAL OUT*/,
+                                           size_t filename_size)
+{
+    memquery_iter_t iter;
+
+    memquery_iterator_start(&iter, NULL, dynamo_heap_initialized);
+
+    print_file(STDERR, "\nfake_memquery_result results_TODORENAME[100] = {");
+    int iters_count = 0;
+    while (memquery_iterator_next(&iter)) {
+        iters_count++;
+        print_file(STDERR, "{\n");
+        print_file(STDERR,
+                   ".iter_result = {"
+                   " .vm_start = (app_pc)" PFX ","
+                   " .vm_end = (app_pc)" PFX ","
+                   " .prot = %X,"
+                   " .comment = \"%s\" },",
+                   iter.vm_start, iter.vm_end, iter.prot, iter.comment);
+
+        app_pc mod_base = 0, mod_end = 0;
+        bool is_header = false;
+
+        if (TEST(MEMPROT_READ, iter.prot)
+            /* We have observed segfaults reading data from very high addresses,
+             * even though their mappings are listed as readable.
+             */
+            && iter.vm_start < (app_pc)0xffffffff00000000 &&
```

Comment:
Actually that clears up a mystery about why the 32-bit-related test cases claimed that none of the sections had headers :) So apparently it compiled but it did not work as intended.

I dropped an ifdef X64 around this since this is to workaround a specific problem I found while debugging our internal repro for i#3387, and I didn't observe it on anything else really. Let me know if you have a better workaround (full disclosure: the high addressed mappings were [vsyscall], so maybe that's the better check).